# Lab 01 My first computational experiment
**Molecular Cell Biology | 15-minute instructor introduction, then 40 minutes of student work**

Create a reusable place for your work, analyze a small DNA sequence, and explore a simple model of protein accumulation. No prior programming is required.

Today you will **predict → run → change one thing → compare → save**. All required code is supplied.

**These are two separate experiments.** The DNA is an invented teaching sequence. The protein model uses invented parameters and does not use the DNA sequence or its GC percentage to calculate protein production.

**Start here:** If your instructor shared a Colab link, choose **File → Save a copy in Drive** before editing. If you downloaded this `.ipynb` file, open [Google Colab](https://colab.research.google.com/), choose **Upload notebook**, and select the file. Use a Google account permitted for your class. Rename your working copy `Lab01_YourCourseAlias.ipynb`. A notebook is opened in Colab, not in Word. The separate Word workbook is the readable companion.


## Setup | Minutes 0–16
### 1. GitHub | Minutes 0–8
Sign in or create a free personal account at [GitHub](https://github.com/). Create a **private** repository named `molecular-cell-biology-lab` with a README. A repository stores project files and their recorded history. Never share passwords or verification codes. At minute 8, tell your instructor about an access problem and continue with an approved partner or demonstration.

### 2. Command line | Minutes 8–13
Windows: open **Windows PowerShell** from Start. Mac: open **Terminal** from Spotlight. These tools normally already exist; do not install software for this lab. Type each line below into that terminal, pressing Enter or Return after each:

```text
cd ~
pwd
mkdir molecular-cell-biology-lab
cd molecular-cell-biology-lab
pwd
ls
```

`cd` changes folders; `~` means your home folder; `pwd` shows the current folder; `mkdir` creates a folder; `ls` lists its contents. If the folder already exists, skip `mkdir` and continue. An empty listing is normal. On a browser-only device, explain an instructor-demonstrated command instead.

### 3. Colab | Minutes 13–16
Open this notebook using the instructions at the top. Use a normal **Python 3 CPU** runtime. No paid plan, GPU, Git installation, local Python installation, or Drive mounting is needed. Click the triangular Run button beside a code cell. Double-click a text cell to edit a response; click outside when finished. Start with the greeting below.

**Where things live:** Your terminal works on your computer. Colab runs Python on a remote computer. GitHub stores the versions you commit. They do **not** automatically synchronize. Do not paste Python into PowerShell or Terminal.

Use the supplied code and your own predictions and interpretations; do not use AI-generated answers for this activity.


### Setup checkpoint — edit this text cell
Course alias (not your student ID): **[type here]**  
Repository link, or account-access issue: **[type here]**  
One terminal command I used and what it did: **[type here]**

Do not include passwords, verification codes, account recovery information, or a full home-directory path. Your instructor will receive your notebook through Canvas even though the repository is private.

### First success | Minute 16
Click the Run button beside the cell below. A sentence should appear beneath the code. The text inside the quotation marks is the message. The `print` instruction displays it.

**Do not paste this Python code into PowerShell or macOS Terminal.**

In [5]:
print("Hello, molecular cell biology!")

Hello, molecular cell biology!


## Step 1 | Analyze DNA | Minutes 16–25
### A. Run a baseline
This synthetic sequence is written 5′ to 3′ using the four standard DNA letters. We will count its length and the percentage of letters that are G or C. We are **not** identifying a real gene or predicting its function.

`dna` is a **variable**: a name attached to a value. The quoted letters are a **string**: text. `len(dna)` counts characters, and `.count("G")` counts uppercase G letters. A line beginning with `#` is a comment for a human reader.

Run the baseline cell. Leave it unchanged afterward.

In [6]:
# Synthetic teaching sequence; use uppercase letters and no spaces.
dna = "ATGAAACCGGCTTTATAA"
length = len(dna)
gc_count = dna.count("G") + dna.count("C")
gc_percent = 100 * gc_count / length

print("Length (bases):", length)
print("G + C (bases):", gc_count)
print("GC (%):", round(gc_percent, 1))

Length (bases): 18
G + C (bases): 6
GC (%): 33.3


### B. Predict, then change one base
In the **second DNA code cell below**, replace only the quoted sequence with:

`ATGAAGCCGGCTTTATAA`

This changes the **sixth base from A to G**, without adding or removing a base. Do not change the baseline cell.

**Before running the changed cell, edit this prediction:**  
I predict that the sequence length will **[stop]**, and the GC percentage will **[increase]**, because **[a base was added without changing the sequence length]**.

Now make the replacement and run the second DNA cell.

In [7]:
# Synthetic teaching sequence; use uppercase letters and no spaces.
dna = "ATGAAGCCGGCTTTATAA"
length = len(dna)
gc_count = dna.count("G") + dna.count("C")
gc_percent = 100 * gc_count / length

print("Length (bases):", length)
print("G + C (bases):", gc_count)
print("GC (%):", round(gc_percent, 1))

Length (bases): 18
G + C (bases): 7
GC (%): 38.9


### C. Compare — edit this text cell
| Measurement | Baseline | One-base change |
|---|---:|---:|
| Length, bases | [18] | [18] |
| G + C, bases | [6] | [7] |
| GC, % | [33.3] | [38.9] |

What changed, and what stayed the same? **[The length stayed the same but the gc increased.]**

Could this GC result alone establish a change in protein function? Why or why not? **[Yes because it will either stop or mutate and niether of which will allow the protein to function correctly.]**

The code counts characters literally. Today, keep the DNA uppercase, nonempty, and free of spaces or other symbols.

## Step 2 | Simulate protein accumulation | Minutes 25–33
### A. Understand the model
Our invented cell starts with **0 arbitrary units of protein**. In each **one-minute step**, it loses **20% of the amount present at the beginning of the step**, and then receives **10 new units**.

The **model** is that rule and its assumptions. The **simulation** is the computer repeatedly applying the rule to calculate amounts over time. The printed values are calculated predictions, **not measurements from real cells**.

The amount is a continuous, relative quantity, not a count of individual molecules. Each step always represents one minute. The loss fraction is fixed at 0.20 per step; it is not a measured rate constant.

**Hand check:** After the first minute, the amount should be 10. After the next minute, 8 of those units remain and 10 new units are added.

The indented lines under `for` repeat. `range(1, 7)` means minutes 1 through 6; the endpoint 7 is excluded. Leave the indentation unchanged. `round` changes only what is displayed, not the unrounded running amount.

In [8]:
# Each loop is one minute. Amounts are arbitrary protein units.
production = 10       # Units added in each one-minute step.
loss_fraction = 0.20  # Fraction of the existing pool lost each step.
protein = 0.0        # Starting amount; reset every time this cell runs.

print("Minute | Protein (arbitrary units)")
print(0, "|", protein)
for minute in range(1, 7):
    protein = protein * (1 - loss_fraction) + production
    print(minute, "|", round(protein, 2))

Minute | Protein (arbitrary units)
0 | 0.0
1 | 10.0
2 | 18.0
3 | 24.4
4 | 29.52
5 | 33.62
6 | 36.89


### B. Predict, then change one parameter
In the **second model code cell below**, change only `production = 10` to `production = 20`. Leave the loss fraction, starting amount, and number of minutes unchanged. A **parameter** is a setting in the model.

**Before running the changed cell, edit this prediction:**  
Compared with the baseline, the amount after six minutes will **[81]**, because **[its a similar distance apart]**.

Make the change and run the second model cell. Each cell starts its own simulation from zero.

In [9]:
# Each loop is one minute. Amounts are arbitrary protein units.
production = 20       # Units added in each one-minute step.
loss_fraction = 0.20  # Fraction of the existing pool lost each step.
protein = 0.0        # Starting amount; reset every time this cell runs.

print("Minute | Protein (arbitrary units)")
print(0, "|", protein)
for minute in range(1, 7):
    protein = protein * (1 - loss_fraction) + production
    print(minute, "|", round(protein, 2))

Minute | Protein (arbitrary units)
0 | 0.0
1 | 20.0
2 | 36.0
3 | 48.8
4 | 59.04
5 | 67.23
6 | 73.79


### C. Compare — edit this text cell
Amount after minute 2 in the baseline: 18 arbitrary units**  
Amount after minute 6 with production = 10: 36.89 arbitrary units**  
Amount after minute 6 with production = 20: 73.79 arbitrary units**

One quantity I kept unchanged was protein. Keeping it unchanged helped because the variable would be different.

One assumption or omitted biological process that limits this model is live timing.

**Optional, only after the required work:** Run the unchanged second model cell again. Do the results change? Explain your observation. Do not change more parameters yet.

## Checks and saving | Minutes 33–40
1. Complete every response marked `[type]` or `[type here]`. Choose **Runtime → Run all**. Confirm that all five code cells run without red errors and that both baseline cells remain unchanged. The second DNA cell should contain the sixth-base A-to-G change; the second model cell should use `production = 20`.
2. Keep your own working copy in Drive. Download it using **File → Download → Download .ipynb** (menu wording may vary). Keep the notebook format and its outputs. Do not choose `.py`, and do not turn on “Omit code cell output when saving this notebook.”
3. In your private GitHub repository, choose **Add file → Upload files**. Upload the completed notebook and commit with the message `Add Lab 01 sequence analysis and protein simulation`. Confirm that the file appears. A **commit** is a recorded version with an explanatory message. This does not create automatic synchronization.
4. Submit the same completed `.ipynb` file to **Canvas**, plus your repository link in the submission comment. Your instructor can grade this file without access to the private repository. Keep a downloaded copy.

**Access issue?** Tell the instructor and record it in your setup checkpoint. Use an approved paired or demonstration route, keep your own predictions and results in the Word workbook, and submit that written record if directed. Do not spend the full lab troubleshooting.


## Next steps
Keep this repository and notebook. In the next lab, you will reopen this work, turn sequence data into a small file, and analyze a new sequence. Later you will compare several sequences, make plots, add biological detail to models, and learn to run and track code locally.

### Software references
[GitHub account setup](https://docs.github.com/en/account-and-profile/how-tos/account-management/creating-an-account-on-github) · [Repository quickstart](https://docs.github.com/en/repositories/creating-and-managing-repositories/quickstart-for-repositories) · [Uploading files](https://docs.github.com/en/repositories/working-with-files/managing-files/adding-a-file-to-a-repository) · [Colab FAQ](https://research.google.com/colaboratory/faq.html) · [Python strings](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str) · [Python loops](https://docs.python.org/3/tutorial/controlflow.html#for-statements)

**Provenance:** Synthetic sequence, model, parameters, and teaching prompts were constructed for this activity; no external biological dataset or experimentally calibrated model is used.